Frame requester and First person controls.

autra opened the pull request:
This PR adds the notion of FrameRequester: a module can subscribe to each request animation frame. Last commit adds a first person controls that uses this feature.

I preferred to make one PR because FPControls gives meaning to this FrameRequester, but I can split it if you want. Maybe a simpler example can be found to illustrate this.

@iTowns/reviewers please r?

autra committed:
Feature: ability to hook a module to requestAnimationFrame

This adds the notion of FrameRequester, something that can request to be
notified at each frame.

autra committed:
Feature: add option to not create GlobeControls

autra committed:
Fix: allow canvas to be focused by default

autra committed:
Feature: add fly controls

autra commented on src/Renderer/c3DEngine.js:
```diff
@@ -38,11 +38,13 @@ function c3DEngine(rendererOrDiv) {
-    //
+    var canvas = document.createElement('canvas');
+    canvas.tabIndex = -1;
+    canvas.addEventListener('click', () => canvas.focus());
```
A bit hacky, but I didn't find better :-)

peppsac replied:
You could add a comment explaining why you did add this?

peppsac replied:
Also, the canvas doesn't need to be created at all if `renderer` is supplied by the user.

autra replied:
The alternative would be to use viewerDiv instead of canvas for these events. WDYT?

peppsac replied:
I have no strong feeling regarding this.

Or maybe you could move this below the `renderer` creation and use `renderer.domElement` instead?

autra replied:
I have a better idea: I think i'll add some options to define the focus policy in my FPControls instead. By default, it'll do nothing and let the client completely deal with focus, but it'll have a `focusOnClick` and `focusOnHover` options.

peppsac commented on src/Core/Prefab/GlobeView.js:
```diff
@@ -84,6 +84,7 @@ export const GLOBE_VIEW_EVENTS = {
  * @param {object=} options - see {@link View}
  */
 function GlobeView(viewerDiv, coordCarto, options) {
+    options = options || {};
```
Maybe use the default parameter syntax for GlobeView constructor instead?

peppsac commented on src/Core/Prefab/GlobeView.js:
```diff
@@ -205,26 +206,9 @@ function GlobeView(viewerDiv, coordCarto, options) {
     const positionTargetCamera = positionCamera.clone();
```
This could be moved in the `if (!options.noControls) {` block I think.

And maybe you should always set the camera position to `coordCarto` - even if the user didn't ask for `GlobeControls` to be used?

peppsac commented on src/Renderer/ThreeExtended/FPControls.js:
```diff
@@ -0,0 +1,144 @@
+    this.view.notifyChange(false);
+}
+
+class FPControls extends THREE.EventDispatcher {
```
What about `FirstPersonControls` name?

peppsac commented on src/Renderer/ThreeExtended/GlobeControls.js:
```diff
@@ -356,22 +350,19 @@ function defer() {
+ * @param {options} options
  */
-function GlobeControls(view, target, domElement, viewerDiv, radius, getPickingPosition) {
+function GlobeControls(view, target, radius, options) {
```
```js
function GlobeControls(view, target, radius, options = {}) {
```
?

peppsac commented on src/Renderer/ThreeExtended/GlobeControls.js:
```diff
@@ -554,7 +543,7 @@ function GlobeControls(view, target, domElement, viewerDiv, radius, getPickingPo
             this.camera.zoom = Math.max(this.minZoom, Math.min(this.maxZoom, this.camera.zoom * dollyScale));
             this.camera.updateProjectionMatrix();
-            this.dispatchEvent(this.changeEvent);
+            this._view.notifyChange(true);
```
What about using: `this._view.notifyChange(true, this.camera);` to indicate to the update cycle that the change was caused by a camera movement? (same comments applies to FPControls obviously)?

peppsac reviewed (commented):
Thanks, this looks useful.

I quickly skimmed through the diff and left a few minor comments.
I'll probably do a follow-up  review next week.

peppsac commented on src/Core/View.js:
```diff
@@ -278,4 +279,49 @@ View.prototype.getLayers = function getLayers(filter) {
 
+/**
+ * @typedef {object} FrameRequester
+ * @property {Function(dt, firtLoop)} update - Method that will be called each
```
typo: `firtLoop` => `firstLoop`. Maybe `updateLoopRestarted` is a better name too?

I'm still wondering if this boolean is really useful..

autra replied:
What I don't like is that it's so easily messed up if controls (for instance) don't plug them well into this system. It relies on FrameRequesters to call `notifyChange` in the same microtask as `MainLoop._step` to work correctly (which GlobeControls does not yet). Otherwise, firstLoop will always be true.

Apart from that, I use it in `FPControls`, so it's at least *a bit* useful...

peppsac replied:
The usage in `FPContols` could be moved in `MainLoop`:
```js
frameRequester.update(this._firstLoop ? 16 : dt);
```
But let's keep it for now. What do you think of my name proposal?

autra replied:
> What do you think of my name proposal?

I agree, ameding now.

> The usage in FPContols could be moved in MainLoop:

I don't think so, because now dt is the *real* interval between 2 requestAnimationFrame. So it's never undefined any more. I think it's less magic and more useful this way.

autra commented:
@peppsac updated. 

For the changeSource of `view.notifyChange`, there are some places where I didn't pass the camera, please check if it makes sense.

autra reviewed (commented):
EDIT: ignore, wrong tab :-)

autra committed:
Fix: correctly update geometry in all cases

When the changeSource is not in the tile tree (for instance
camera), it means we need to update from the root, not that we don't
need to update at all.

autra commented:
@peppsac updated

peppsac reviewed (approved):
LGTM :)

gchoqueux reviewed (dismissed):
there's visibilty bug in globe

autra commented:
> there's visibilty bug in globe

As usual, I need you to be a bit more specific please :-)

On a more general basis, please be aware that this PR will uncover bugs, not necessarily directly caused by it, but just because GlobeView is not made to use anything else than GlobeControls right now. Can you reproduce using index.html?

gchoqueux commented:
I made a quick remark to alert, I'll come back soon

gchoqueux commented:
> Visibilty bug in globe

Means that there are tiles on the globe that are displayed or hidden when they should not

It is very simple to reproduce this bug:
zoom and de-zoom quickly when the camera pyramid intersects tiles

Warning: an invisible bug in the master isn't a bug

gchoqueux commented:
There's also an error message in `GlobeControls.updateCamera` :
Uncaught TypeError: Can not read property 'clientWidth' of undefined

peppsac commented:
@gchoqueux : you can be more helpful by following simple steps (here and in all your reviews):
- avoid 1 line comments ('it doesn't work')
- always specify steps to reproduce, or at the very least include a description of the issue (instead of waiting for the PR owner to ask for them)
- you can also do some bit of investigation. You're the top contributor of itowns, so you can probably guess sometimes where the culprit code is. E.g: 'if I remove these 2 lines from your changeset the bug disappear' etc

peppsac commented:
@gchoqueux @autra : I think the issue is caused by the cleanup of GlobeControls.

You replaced:
```js
-    this.controls.addEventListener('change', () => {
-        this.camera.update();
-        this.notifyChange(true);
-    });
```
(see  https://github.com/iTowns/itowns/pull/392/files#diff-caf5227be4f9e2f283d82df32a678e4bL220 )
with direct calls to `view.notifyChange()` in `GlobeControls`.

This is a good thing, but it uncovered a bug in `MainLoop`: the camera is updated *after* the step() function and *before* the render function.

I think a good fix would be to update the cameraat the beginning of step: this way the update will proceed with the camera at the correct position.

Moving this block at the beginning of `step` should be enough:
```js
    const dim = this.gfxEngine.getWindowSize();
    view.camera.resize(dim.x, dim.y);
    view.camera.update();
```

autra commented:
> Warning: an invisible bug in the master isn't a bug

I disagree on this. Will we close geoportail (or others, for that matter) bugs when we cannot reproduce on one of the html in `examples/`? Anyway, does not matter so much here, as I can reproduce it on master. So not a bug on this PR.

autra commented:
Beside, I'm not even sure what the bug is because I still don't have a gif or a video showing it.

Let me do some guesswork. Is it the blue triangle on top of the screen in this gif?

![bug_maybe](https://user-images.githubusercontent.com/1576598/28316999-69d7f5cc-6bc5-11e7-8f11-67a8fe52f23c.gif)

Because if so, this is a clean master.

peppsac commented:
Your gif looks seriously like issue #87 
Could you display bounding-boxes to be sure?

autra commented:
My gif indeed looks like #87.

autra committed:
Fix: update the camera before updating layers

If we do not, there can be a 1 frame position difference for the
calculus of whether or not a tile is visible. When moving quickly (but
not *that* quickly), we have large portion of the geometry not
immediately visible.

autra commented:
Added one commit fixing (hopefully) @gchoqueux's issue. Also amended `Feature: add first person controls` with a correct camera update in `FirstPersonControls.update` for the same reason. Thanks @peppsac for the explanations.

@peppsac @gchoqueux WDYT?

peppsac commented on src/Core/Prefab/GlobeView.js:
```diff
@@ -163,7 +163,7 @@ function GlobeView(viewerDiv, coordCarto, options) {
             return [commonAncestor];
         } else {
-            return [];
+            return layer.level0Nodes;
```
Nice. But the purpose is defeated by https://github.com/iTowns/itowns/pull/392/files#diff-caf5227be4f9e2f283d82df32a678e4bL150

The above loop assumed that `changeSources` content were all `TileMesh`.
With this PR this is no longer the case (which is a good thing), but you need add a safeguard in the loop for this case.

Options:
- add `if (source.isCamera) { ... }` but this will break as soon as a non camera object is used
- add a `isTileMesh` property ?
- test for `isMesh` and `node.level` existence?

autra replied:
I see. It seems that L158 saves me a little though (when sources does not have materials).

Actually, I think the best safeguard would be to test `source.layer === 'globe'`, because of line 137. WDYT?

autra commented:
Ok updated again. Please look especially at 9b7b96d.

autra commented:
Updated again.

peppsac commented:
LGTM now

autra commented:
@gchoqueux WDYT?

gchoqueux commented:
I'll do the review as soon as possible.

Otherwise first person should have a notion of gravity?

autra commented:
> Otherwise first person should have a notion of gravity?

Actually, maybe a better name for my control would be `FlyControls`.

autra commented:
To elaborate on this. I *do* have another controls that have a notion of verticality (if that's what you meant by "gravity"). At the moment, it works only for planar view (or other scene that assumes a planar world, like local pointcloud for instance). I need to make it able to change its vertical axe for the globe. I'll submit it later when this is done.

autra commented:
Just pushed a commit that renames it to `FlyControls`. I can squash it to the original commit if you want.

autra commented:
Ok so the commits are tidy and commit messages should be ok now. I've added a help div on the 2 example pages, renamed and changed title of globe_fp.html (now globe_fly.html

autra commented:
Updated: the commits are tidy and commit messages should be ok now. I've added a help div on the 2 example pages, renamed and changed title of globe_fp.html (now globe_fly.html).

peppsac commented:
Thanks for all the changes @autra 
Now it's time to move forward: I'm merging this PR.

gchoqueux commented:
Why did you merge before my review?

https://github.com/iTowns/itowns/pull/392#issuecomment-316155803

peppsac commented:
Because @autra answered all your comments and I have other branches waiting on this one to be merged.

And I didn't know when you'll be available again to review (knowing that you're not working 100% on iTowns currently). 

Feel free to contact me privately to talk about this.

Meanwhile we have some PR waiting for a review: #397, #396, #386 and #309

gchoqueux commented:
This https://github.com/iTowns/itowns/pull/392#issuecomment-316018782 has been disregarded. 
This bug is in the master when you resize the windows, but it isn't very important. On the other hand, it shows that to merge quickly and rejecting the review isn't the best method.

gchoqueux commented:
And if you want to know when I'm available : Ask me ;)

peppsac commented:
Where are the details of the issue you reported in this comment?
Sorry but it's again a "something is broken" comment - not helpful and easy to forget.

> And if you want to know when I'm available : Ask me ;)

Sure. But as you weren't on IRC we assumed we weren't working today.

gchoqueux commented:
> Where are the details of the issue you reported in this comment?

I guess if you read this review, you would have asked for details.

**There's also an error message in GlobeControls.updateCamera :
Uncaught TypeError: Can not read property 'clientWidth' of undefined**

Isn't detailed enough? What do not you understand?
This  `GlobeControls.updateCamera` has been modified in this PR?

> I didn't know when you'll be available again to review

Sorry I thought you did not know, So ask the question about when I'll be available

gchoqueux commented:
If you want to accelerate a PR, it's more appropriate to use this method : https://github.com/iTowns/itowns/pull/221#issuecomment-285635382

autra commented:
@gchoqueux 

> There's also an error message in GlobeControls.updateCamera :
Uncaught TypeError: Can not read property 'clientWidth' of undefined

I believe this is fixed (I tested both on chrome and Firefox). Please let me know if you still encounter this issue, I'll make a follow-up PR.

> Sorry I thought you did not know, So ask the question about when I'll be available

No, that's not how it works. It's not our responsibility to organize our work around your schedule. If you're off, you're off. If you're working, you're working. And if you're working, you are also connected to IRC, please (where you can ask us to wait if you want).

But you can't just come, put a "change requested" with a "wait for my review" like you did in other PR, then we have to wait for you to come back before anything can move. When @peppsac or I are on PTO, on other leave or on other projects, we don't request you to wait for us. Does this make sense to you?

> If you want to accelerate a PR, it's more appropriate to use this method : #221 (comment)

And I didn't put a "wait for my review" there because I wasn't working on itowns at that point, even though I wouldn't have merged this.

Here, @peppsac only chose to merge because your review didn't express a disagreement in the principle of it. In short, you said nothing that can't be addressed in minor follow-up PRs. Again feel free to open issues or ping me on IRC if you want such follow-ups.

autra commented:
To follow-up

> This #392 (comment) has been disregarded.

I believe it hasn't, but I'm double-checking right now.

autra commented:
To follow-up again @gchoqueux, and to elaborate on why you should always give *precise* step to reproduce:

There was actually 2 situations where this error fired: 
- one at startup, that is fixed
- one at resize, still there indeed

You were speaking about the one when resizing, but only gave the information it was so *yesterday*. So :

> There's also an error message in GlobeControls.updateCamera :
> Uncaught TypeError: Can not read property 'clientWidth' of undefined
> 
> Isn't detailed enough? What do not you understand?

No, an error in the console is *not* "detailed enough". It's useless without steps to reproduce. And also:

> I guess if you read this review, you would have asked for details.

That's again not how it works. We are not here to do a guessing game. You need to provide the details, and especially the steps to reproduce. This will guarantee a better experience for your reviewee.

A must-read in bug reporting https://developer.mozilla.org/en-US/docs/Mozilla/QA/Bug_writing_guidelines#Writing_a_clear_summary

And I quote:

> Steps to reproduce are the **most important part of any bug report**. If a developer is able to reproduce the bug, the bug is very likely to be fixed. If the steps are unclear, it might not even be possible to know whether the bug has been fixed.
